### ISSUE TITLE
RF: BUF_FULL - disables the RF part on prop-mode (CC1310)

### ISSUE THREAD

**Issue by serdarkose** [contiki-ng/contiki-ng#778]
The issue is related to https://github.com/contiki-os/contiki/pull/2161
----------------------------------------------------------------------------------------
# Issue 

As the above issue states, `RF: BUF_FULL` occurs due to mid or high volume RF traffic. However, it may also occur at a single node network if the node consume resources, especially at the start-up.

Once the problem occurs, its RF part can't recover. It turns of RX and partly turns on the TX (to transmit DIS), otherwise, RF part doesn't work.

# Brief set-up details:
**HW:**
- I have a border router: RPI 2 + CC1310 launchpad
- one CC1310 board that runs the modified examples/hello-world project (code is at the end)

**Routine:**
1. Border router is up and running
2. open a new terminal on the RPI and start ping6'ing to the node address
3. power the node..  it immediately outputs `RF:BUF_FULL` error at the start-up. After that RF part doesn't work properly. As I experience, RF part is not recovered once the error is occurred.

# As I observe:
 `cc26xx_rf_cpe1_isr()` is called after the `IRQ_RX_BUF_FULL` is asserted. The ISR calls [process_poll(&rf_core_process)](https://github.com/contiki-ng/contiki-ng/blob/4d6c8898c27440413a6a7e600053bf68b699e2a6/arch/cpu/cc26x0-cc13x0/rf-core/rf-core.c#L737) which relies on `read_frame()` will empty the buffer. However, this method is too slow to empty the buffer. Before it happens, the RF part is already corrupted or works wrong.  It also happens that the RF part may receive new packets before it frees the buffer.

Another problem I observe is that  it turns off the RF core where it is supposed to be powered on: 
 ` read_frame() ` calls  [rf_core_convert_rat_to_rtimer(rat_timestamp) ](https://github.com/contiki-ng/contiki-ng/blob/4d6c8898c27440413a6a7e600053bf68b699e2a6/arch/cpu/cc26x0-cc13x0/rf-core/prop-mode.c#L867) it flows to [line 663](https://github.com/contiki-ng/contiki-ng/blob/4d6c8898c27440413a6a7e600053bf68b699e2a6/arch/cpu/cc26x0-cc13x0/rf-core/rf-core.c#L663) , it uses `rf_is_on()` to read `PDSTAT1RFC` register's value if the RF core is 1 (ON) or 0 (Off). 
 The register value is **1** and ` smartrf_settings_cmd_prop_rx_adv.status` has `0x3801 ` value (RF BUFFER FULL ERR) at that specific processing time. 

the verification on [return smartrf_settings_cmd_prop_rx_adv.status == RF_CORE_RADIO_OP_STATUS_ACTIVE ](https://github.com/contiki-ng/contiki-ng/blob/4d6c8898c27440413a6a7e600053bf68b699e2a6/arch/cpu/cc26x0-cc13x0/rf-core/prop-mode.c#L253)  returns  `false` to the [line 272](https://github.com/contiki-ng/contiki-ng/blob/4d6c8898c27440413a6a7e600053bf68b699e2a6/arch/cpu/cc26x0-cc13x0/rf-core/prop-mode.c#L272) as it only compares with 0x0002.  This misleads code to assume that the rf core was off earlier and should be turned off before exiting the function. However, the rf core should be kept on, otherwise, when RF core is turned off in that case, the RX will be never turned on again. In some cases, it turns on the RF core to transmit DIS and turns off again.  

However, it is not a pure RF core power down issue, because I manually commented out inside [off()](https://github.com/contiki-ng/contiki-ng/blob/4d6c8898c27440413a6a7e600053bf68b699e2a6/arch/cpu/cc26x0-cc13x0/rf-core/prop-mode.c#L1072)  function that prevents RF core to be powered off. It keeps the RF core to be powered on, but the radio works terrible, as bad as it is turned off. 

As an example due to `#define LOG_CONF_LEVEL_FRAMER 		LOG_LEVEL_DBG` printf outputs many "RF out" message on the serial output,, but wireshark/sensniff shows that only very few of the "RF out" are actually transmitted. It is the same for the "RF in" packets, it can receive very few of the actually transmitted packets by the other nodes. 

So far, I think there are two main issues to fix:
1. prevent it to false-power off, maybe remove the verification on [line 253](https://github.com/contiki-ng/contiki-ng/blob/4d6c8898c27440413a6a7e600053bf68b699e2a6/arch/cpu/cc26x0-cc13x0/rf-core/prop-mode.c#L253) or replace it as ` return ( (smartrf_settings_cmd_prop_rx_adv.status == RF_CORE_RADIO_OP_STATUS_ACTIVE)  || (smartrf_settings_cmd_prop_rx_adv.status  == RF_CORE_RADIO_OP_STATUS_PROP_ERROR_RXBUF ) ) ;` (?)
2. I guess the RF state corrupts or does no longer holds the RF's state accurately once the RF_BUF occurs/following executions. I am not sure what is getting broken, but is it possible to immediately solve the RF: BUFF_FULL issue inside the interrupt instead of [line 737](https://github.com/contiki-ng/contiki-ng/blob/4d6c8898c27440413a6a7e600053bf68b699e2a6/arch/cpu/cc26x0-cc13x0/rf-core/rf-core.c#L737) that I think it is too slow to act,  OR add 'reset()' function for the entire RF (or wherever it has the issue) to recover it properly (?)

- Another question, once the RF_Buffer interrupt is asserted, shall the interrupt disabled till it is solved? The current implementation takes too long to clear the rf_buf and the same interrupt occurs sometime before the buffer is cleared.

Last question about the driver is that there are many `RF_CORE_CMD_ERROR` controls, but some functions are not handling the error cases more than PRINTF. Would it be better to add error handling wherever RF_CORE_CMD_ERROR may be returned? As an example:
BUF_FULL interrupt calls read_frame(), and at the end, it calls release_data_entry(), due to `rf_core_rx_is_full` variable, [it calls rx_on_prop()](https://github.com/contiki-ng/contiki-ng/blob/4d6c8898c27440413a6a7e600053bf68b699e2a6/arch/cpu/cc26x0-cc13x0/rf-core/prop-mode.c#L806) that function calls `rf_cmd_prop_rx()` and it gets error   [ if(cmd_rx_adv->status != RF_CORE_RADIO_OP_STATUS_ACTIVE) {](https://github.com/contiki-ng/contiki-ng/blob/4d6c8898c27440413a6a7e600053bf68b699e2a6/arch/cpu/cc26x0-cc13x0/rf-core/prop-mode.c#L453) then it doesn't do any error handling where it assumes that the RX is set, but since the RX couldn't set it up, then RX part doesn't work any longer, 

#  Code part
----------------------------------------------------------------------------------------------
hello-world.c
```
#include "contiki.h"
#include "project-conf.h"
#include <stdio.h> /* For printf() */

static void add_100ms_delay( void ){
      volatile unsigned long ulLoopCount = 0;
      for(ulLoopCount = 442000; ulLoopCount > 0; ulLoopCount--) 
      {
 	    watchdog_periodic();   
      }
}

/*---------------------------------------------------------------------------*/
PROCESS(hello_world_process, "Hello world process");
AUTOSTART_PROCESSES(&hello_world_process);
/*---------------------------------------------------------------------------*/
PROCESS_THREAD(hello_world_process, ev, data)
{
  static struct etimer timer;
  PROCESS_BEGIN();
  /* Setup a periodic timer that expires after 1 seconds. */
  etimer_set(&timer, CLOCK_SECOND * 1);

  while(1) {
	uint8_t i;
	add_100ms_delay();
	add_100ms_delay();             //The RF: ERROR already occurs before reaching here
	for(i=0; i <50; i++){
		printf("%"PRIu8"\t", i);
	}

    /* Wait for the periodic timer to expire and then restart the timer. */
    PROCESS_WAIT_EVENT_UNTIL(etimer_expired(&timer));
    etimer_reset(&timer);
  }

  PROCESS_END();
}
```

project-conf.h includes:
```
#define LOG_CONF_LEVEL_MAIN 		LOG_LEVEL_DBG
#define LOG_CONF_LEVEL_RPL 		LOG_LEVEL_NONE
#define LOG_CONF_LEVEL_TCPIP 		LOG_LEVEL_NONE
#define LOG_CONF_LEVEL_IPV6 		LOG_LEVEL_NONE
#define LOG_CONF_LEVEL_6LOWPAN 		LOG_LEVEL_NONE
#define LOG_CONF_LEVEL_NULLNET 		LOG_LEVEL_NONE
#define LOG_CONF_LEVEL_MAC 		LOG_LEVEL_NONE
#define LOG_CONF_LEVEL_FRAMER 		LOG_LEVEL_DBG
#define LOG_CONF_LEVEL_6TOP 		LOG_LEVEL_NONE
#define LOG_CONF_LEVEL_COAP 		LOG_LEVEL_NONE
#define LOG_CONF_LEVEL_LWM2M 		LOG_LEVEL_NONE

#define NBR_TABLE_CONF_MAX_NEIGHBORS 20
#define CSMA_CONF_MAX_FRAME_RETRIES 7
#define QUEUEBUF_CONF_NUM 8
#define RPL_MRHOF_CONF_SQUARED_ETX 0
#define RPL_CONF_PROBING_INTERVAL (90 * CLOCK_SECOND)
#define RPL_CONF_DIS_INTERVAL (30 * CLOCK_SECOND)
#define RPL_CONF_DIO_INTERVAL_MIN        12
#define RPL_CONF_DIO_INTERVAL_DOUBLINGS  (20-RPL_CONF_DIO_INTERVAL_MIN)
#define RPL_CONF_DEFAULT_LIFETIME 30
#define RPL_CONF_MAX_RANKINC 0

#define RF_CORE_CONF_CHANNEL      26
#undef IEEE802154_CONF_DEFAULT_CHANNEL
#define IEEE802154_CONF_DEFAULT_CHANNEL RF_CORE_CONF_CHANNEL

#define ROM_BOOTLOADER_ENABLE      1

//#undef LPM_MODE_MAX_SUPPORTED
//#define LPM_MODE_MAX_SUPPORTED     LPM_MODE_AWAKE
```

Makefile
```
CONTIKI_PROJECT = hello-world
TARGET = cc26x0-cc13x0
BOARD = srf06/cc13x0

all: $(CONTIKI_PROJECT)

CONTIKI = ../..
include $(CONTIKI)/Makefile.include
```

Below is the serial output. The error occurs at the very beginning. My guess is that it happens like this: the start-up routine calls netstack_init(), it calls
```
  NETSTACK_RADIO.init();
  NETSTACK_MAC.init();
  NETSTACK_NETWORK.init();
```
**radio.init() powers the RF core and RX. I guess it receives the packet before mac and/or network init are completed which causes the problem.**  Can it be the case? Is it possible to discard everything till the whole RF is set up completely?

```
[INFO: Main      ] Starting Contiki-NG-d05faed-dirty
[INFO: Main      ] - Routing: RPL Lite
[INFO: Main      ] - Net: sicslowpan
[INFO: Main      ] - MAC: CSMA
[INFO: Main      ] - 802.15.4 PANID: 0xd48a
[INFO: Main      ] - 802.15.4 Default channel: 26
[INFO: Main      ] Node ID: 16262
[INFO: Main      ] Link-layer address: 0012.4b00.197b.3f86
[INFO: Main      ] Tentative link-local IPv6 address: fe80::212:4b00:197b:3f86
[DBG : CC26xx/CC13xx] With DriverLib v0.47020
[INFO: CC26xx/CC13xx] TI SmartRF06EB + CC13xx EM
[DBG : CC26xx/CC13xx] IEEE 802.15.4: No, Sub-GHz: Yes, BLE: No, Prop: No
[INFO: CC26xx/CC13xx]  RF: Channel 26
RF Error

RF: BUF_FULL

0       1       2       3       4       5       6       7       8       9       10      11  !
rf_core_send_cmd: RF was off
rf_cmd_prop_rx: send_cmd ret=0, CMDSTA=0x0000081f, status=0x0000
[INFO: Frame 15.4] In:  1 0012.4b00.13f1.e0b9 0012.4b00.197b.3f86 21 88 (109)
0       1       2       3       4       5       6       7       8       9       10      11  4
0       1       2       3       4       5       6       7       8       9       10      11  4
0       1       2       3       4       5   
```

# EDIT: a possible solution - but probably not complete
Below changes fixes the issues but RF performance is decreased as I observe:
 
1. replace  [line 253](https://github.com/contiki-ng/contiki-ng/blob/4d6c8898c27440413a6a7e600053bf68b699e2a6/arch/cpu/cc26x0-cc13x0/rf-core/prop-mode.c#L253) with
`return ( (smartrf_settings_cmd_prop_rx_adv.status == RF_CORE_RADIO_OP_STATUS_ACTIVE) || (smartrf_settings_cmd_prop_rx_adv.status == RF_CORE_RADIO_OP_STATUS_PROP_ERROR_RXBUF ) ) 
`
This will prevent the wrong RF  power off action

2.  replace `static` with `extern` on [release_data_entry()](https://github.com/contiki-ng/contiki-ng/blob/4d6c8898c27440413a6a7e600053bf68b699e2a6/arch/cpu/cc26x0-cc13x0/rf-core/prop-mode.c#L787) 
The purpose is for `rf-core.c` to be able to call/access it once the `RF: BUF_FULL` occurs. It'd take an immediate action to clear the buffer.

Add the `release_data_entry()` instead of [ process_poll(&rf_core_process)](https://github.com/contiki-ng/contiki-ng/blob/4d6c8898c27440413a6a7e600053bf68b699e2a6/arch/cpu/cc26x0-cc13x0/rf-core/rf-core.c#L737)  (comment it out)
I'd assume proper implementation would be to define an `extern void clear_rx_buffer()` to call the `release_data_entry()`

3. replace   [if(rf_is_on()) ](https://github.com/contiki-ng/contiki-ng/blob/4d6c8898c27440413a6a7e600053bf68b699e2a6/arch/cpu/cc26x0-cc13x0/rf-core/prop-mode.c#L485) with `if(rf_is_on() && !(smartrf_settings_cmd_prop_rx_adv.status == RF_CORE_RADIO_OP_STATUS_PROP_ERROR_RXBUF ) ) {` 
the purpose is that the code should flow to `rf_cmd_prop_rx()` that is needed to set the RX parameters. But, I am not what parameter are really needed to be fixed or if this function fixes it immediately.

After the above implementations, the code can recover itself after `RF_BUF` occurs, and works OK. However, the RF performance is decreased till it is recovered which is natural, but maybe there is a faster way to recover it that would improve the RF performance.  it'd be great someone who has good knowledge (like George) may suggest a proper fix. 

Here is my ping output, normally the ping replies are about 300-600 ms range, below is after the RF_BUF event,  Wherever the reply takes too long, RF_BUF recovery happens.

I've tested with my own application code where it uses a lot of resources. I tested together with 40 nodes that does burst traffic all together, the node still works fine (after the burst is over).
```
ytes from fd00::212:4b00:197b:3f86: icmp_seq=12775 ttl=63 time=573 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12776 ttl=63 time=1658 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12777 ttl=63 time=1421 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12778 ttl=63 time=1866 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12779 ttl=63 time=935 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12780 ttl=63 time=207 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12781 ttl=63 time=591 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12782 ttl=63 time=434 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12783 ttl=63 time=769 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12784 ttl=63 time=775 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12785 ttl=63 time=939 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12786 ttl=63 time=547 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12787 ttl=63 time=936 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12788 ttl=63 time=710 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12789 ttl=63 time=148 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12790 ttl=63 time=398 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12791 ttl=63 time=935 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12792 ttl=63 time=815 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12793 ttl=63 time=676 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12794 ttl=63 time=1653 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12795 ttl=63 time=661 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12796 ttl=63 time=132 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12797 ttl=63 time=657 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12797 ttl=63 time=968 ms (DUP!)
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12798 ttl=63 time=424 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12799 ttl=63 time=579 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12800 ttl=63 time=743 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12801 ttl=63 time=201 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12802 ttl=63 time=241 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12803 ttl=63 time=693 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12804 ttl=63 time=153 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12805 ttl=63 time=731 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12806 ttl=63 time=152 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12807 ttl=63 time=594 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12808 ttl=63 time=494 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12809 ttl=63 time=734 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12810 ttl=63 time=171 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12811 ttl=63 time=1006 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12812 ttl=63 time=483 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12813 ttl=63 time=928 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12814 ttl=63 time=649 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12815 ttl=63 time=546 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12816 ttl=63 time=1683 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12817 ttl=63 time=814 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12818 ttl=63 time=594 ms
64 bytes from fd00::212:4b00:197b:3f86: icmp_seq=12819 ttl=63 time=734 ms
```

**Comment by MartinNPN** [comment]
Hello,

i had a similar problem, but can't find the reason. Your description looks plausible. The hard delay is something which should be prevented with contiki os, but sometimes it can be necessary or the program itself must be run something a longer time (flash access or something).
The RF driver should work even if the main program blocks, maybe it could loose some data but it shouldn't break. 

I will test your solution and give some feedback, but i even hope someone with more experience will look at your proposal. 


Regards,
Martin

**Comment by MartinNPN** [comment]
the serial output is maybe a problem because the uart uses polling for tx, i created an issue #812 . But this would not fix this issue, cause the reason is the not proper rx buff full handling.

**Comment by serdarkose** [comment]
I think that would improve the behavior as the serial output affects it as it slows down the processing time (due to hard block). but the problem still occurs if the serial output is disabled completely, maybe it has something to do with the rf buffer. I haven't understood if RF can have its own buffer and can drop the packets if there is no space instead of creating error?

**Comment by lafka** [comment]
I'm a bit unclear on the purpose of `rf_is_on/0`. Should it be used to check that the device is in RX mode? the status of the last RX command? or if the RF core is powered up? From code it's definitely not clear... functions like `transmit/1` and `get_rssi/0` uses it for powering up/down the rf core; `rx_on_prop/0` seems to be using it pick the radio operation. Anyone more familiar  with the drive that has any idea?

By removing `rf_is_on/0`, keeping track of the  _intended_ state of the radio, and relying on `rf_core_is_accessible()` (or similar) to check the power state of the device the radio seems to handle the situation alot better (almost immediate recovery after a full RXBUF)

I've made a small proof of concept ( https://github.com/lafka/contiki-ng/commit/7347ff3cdac360753705136f64a90209f94fbfd1 ) which at first glance fixes the mentioned scenario.

Using `rf_core_is_accessible/0` is probably the wrong function here. I was thinking about using `PDSTAT0:RFC_ON` but that does seem to return "maybe powered down" which is not good enough. Any suggestions on which register holds state about the Rf core power would be great.

@serdarkose If rx buffer is full new packets will not be handled.

**Comment by lafka** [comment]
This can be closed now that https://github.com/contiki-ng/contiki-ng/issues/1089 is merged.

**Comment by arurke** [comment]
Presumably resolved by #1089
